src/arc_spice/variational_pipelines/RTC_variational_pipeline.py

```diff
@@ -0,0 +1,545 @@
+        ]
+        # join these to create the full translation
+        full_translation = ("").join(sentence_translations)
+        # get softmax of the logits to get token probabilities
+        softmax_logits = softmax(translator_outputs[0]["raw_outputs"]["logits"], dim=-1)
+        max_token_scores = torch.max(softmax_logits, dim=-1).values.squeeze(dim=0)
+        # record the output and token probabilities
+        confidence_metrics = [
+            {
+                "outputs": translator_output["translation_text"],
+                "probs": max_token_scores,
+            }
+            for translator_output in translator_outputs
+        ]
+        # group metrics under a single dict key
+        stacked_conf_metrics = self.stack_translator_sentence_metrics(
+            confidence_metrics
+        )
+        # add full output to the output dict
+        outputs = {"full_output": full_translation}
+        outputs.update(stacked_conf_metrics)
+        # {full translation, sentence translations, logits, semantic embeddings}
+        return outputs
+
+    def classify_topic(self, text: str) -> dict[str:str]:
+        """
+        Runs the classification model
+
+        Returns:
+            Dictionary of classification outputs, namely the output scores.
+        """
+        forward = self.classifier(text, self.topic_labels)
+        return {"scores": forward["scores"]}
+
+    def stack_translator_sentence_metrics(
+        self, all_sentence_metrics: list[dict[str:Any]]
+    ) -> dict[str : list[Any]]:
+        """
+        Stacks values from dictionary list into lists under a single key
+
+        Returns:
+            Dictionary with each metric under a single key
+        """
+        stacked = {}
+        # we skip the full output
+        for metric in self.naive_outputs["translation"][1:]:
+            stacked[metric] = [
+                sentence_metrics[metric] for sentence_metrics in all_sentence_metrics
+            ]
+        return stacked
+
+    def stack_variational_outputs(self):
+        """
+        Similar to above but this stacks variational output dictinaries into lists
+        under a single key.
+        """
+        # Create new dict
+        new_var_dict = {}
+        # For each key create a new dict
+        for step in self.var_output.keys():
+            new_var_dict[step] = {}
+            # for each metric in a clean inference run (naive_ouputs)
+            for metric in self.naive_outputs[step]:
+                # create a list for each item
+                new_values = [None] * self.n_variational_runs
+                # populate list with items and assign it in new dictionary
+                for run in range(self.n_variational_runs):
+                    new_values[run] = self.var_output[step][run][metric]
+                new_var_dict[step][metric] = new_values
+        # overwrite the existing output dictionary
+        self.var_output = new_var_dict
```

Comment:
why not return?

Reply:
same for other fns

Reply:
No particular reason, we could maybe refactor this in a separate issue? Or I can change this before we merge?

Reply:
will do this now

Reply:
Done